### ISSUE TITLE
Problem converting tensorrt engine after converting yolov8-worldv2 to onnx

### ISSUE THREAD

**Issue by changsubi** [ultralytics/ultralytics#13171]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Bug

ERROR: [TRT]: 4: [shapeCompiler.cpp::evaluateShapeChecks::1180] Error Code 4: Internal Error (kOPT values for profile 0 violate shape constraints: IShuffleLayer /0/model.12/attn/Reshape: reshaping failed for tensor: /0/model.12/attn/gl/Add_output_0 Reshape dimension of -1 has no solution.)

### Environment

```
import os
import sys
import argparse
import warnings
import onnx
import torch
import torch.nn as nn
from copy import deepcopy
from ultralytics import YOLOWorld
from ultralytics import YOLO
from ultralytics.utils.torch_utils import select_device
from ultralytics.nn.modules import C2f, Detect, RTDETRDecoder, WorldDetect, C2fAttn


class DeepStreamOutput(nn.Module):
    def __init__(self):
        super().__init__()

    def forward(self, x):
        x = x.transpose(1, 2)
        boxes = x[:, :, :4]
        scores, classes = torch.max(x[:, :, 4:], 2, keepdim=True)
        classes = classes.float()
        return boxes, scores, classes


def suppress_warnings():
    warnings.filterwarnings('ignore', category=torch.jit.TracerWarning)
    warnings.filterwarnings('ignore', category=UserWarning)
    warnings.filterwarnings('ignore', category=DeprecationWarning)


def yolov8_export(weights, device):
    model = YOLOWorld(weights)
    # model = YOLO(weights)
    model = deepcopy(model.model).to(device)
    for p in model.parameters():
        p.requires_grad = False
    model.eval()
    model.float()
    model = model.fuse()
    for k, m in model.named_modules():
        if isinstance(m, (Detect, RTDETRDecoder, WorldDetect)):
            m.dynamic = False
            m.export = True
            m.format = 'onnx'
        elif isinstance(m, (C2f,C2fAttn)):
            m.forward = m.forward_split
    return model


def main(args):
    suppress_warnings()

    print('\nStarting: %s' % args.weights)

    print('Opening YOLOv8 model\n')

    device = select_device('cpu')
    model = yolov8_export(args.weights, device)

    if len(model.names.keys()) > 0:
        print('\nCreating labels.txt file')
        f = open('labels.txt', 'w')
        for name in model.names.values():
            f.write(name + '\n')
        f.close()

    model = nn.Sequential(model, DeepStreamOutput())

    img_size = args.size * 2 if len(args.size) == 1 else args.size

    onnx_input_im = torch.zeros(args.batch, 3, *img_size).to(device)
    onnx_output_file = os.path.basename(args.weights).split('.pt')[0] + '.onnx'

    dynamic_axes = {
        'input': {
            0: 'batch'
        },
        'boxes': {
            0: 'batch'
        },
        'scores': {
            0: 'batch'
        },
        'classes': {
            0: 'batch'
        }
    }

    print('\nExporting the model to ONNX')
    torch.onnx.export(model, onnx_input_im, onnx_output_file, verbose=False, opset_version=args.opset,
                      do_constant_folding=True, input_names=['input'], output_names=['boxes', 'scores', 'classes'],
                      dynamic_axes=dynamic_axes if args.dynamic else None)

    if args.simplify:
        print('Simplifying the ONNX model')
        import onnxsim
        model_onnx = onnx.load(onnx_output_file)
        model_onnx, _ = onnxsim.simplify(model_onnx)
        onnx.save(model_onnx, onnx_output_file)

    print('Done: %s\n' % onnx_output_file)


def parse_args():
    parser = argparse.ArgumentParser(description='DeepStream YOLOv8 conversion')
    parser.add_argument('-w', '--weights', required=True, help='Input weights (.pt) file path (required)')
    parser.add_argument('-s', '--size', nargs='+', type=int, default=[640], help='Inference size [H,W] (default [640])')
    parser.add_argument('--opset', type=int, default=16, help='ONNX opset version')
    parser.add_argument('--simplify', action='store_true', help='ONNX simplify model')
    parser.add_argument('--dynamic', action='store_true', help='Dynamic batch-size')
    parser.add_argument('--batch', type=int, default=1, help='Static batch-size')
    args = parser.parse_args()
    if not os.path.isfile(args.weights):
        raise SystemExit('Invalid weights file')
    if args.dynamic and args.batch > 1:
        raise SystemExit('Cannot set dynamic batch-size and static batch-size at same time')
    return args


if __name__ == '__main__':
    args = parse_args()
    sys.exit(main(args))

```
I converted onnx to dynamic batch with the corresponding code and then converted onnx to engine again.
but i have some error
ERROR: [TRT]: 4: [shapeCompiler.cpp::evaluateShapeChecks::1180] Error Code 4: Internal Error (kOPT values for profile 0 violate shape constraints: IShuffleLayer /0/model.12/attn/Reshape: reshaping failed for tensor: /0/model.12/attn/gl/Add_output_0 Reshape dimension of -1 has no solution.)

### Are you willing to submit a PR?

- [X] Yes I'd like to help by submitting a PR!

**Comment by changsubi** [comment]
Dynamic batching is not possible, but static batching is possible. And when converting tensorrt, the workspace size must be set high. Thank you for answer.

**Comment by glenn-jocher** [comment]
Hi there,

Thank you for the update and for sharing your findings! It's great to hear that static batching is working for your scenario. Regarding the workspace size for TensorRT conversion, indeed, allocating a larger workspace can often help accommodate the memory demands during the optimization process, especially for complex models.

If you encounter any more issues or have further questions, feel free to reach out. Happy coding! 🚀

**Comment by SunXuan90** [comment]
Any insight why is this happening and possible workaround? I deployed this model to Triton Inference Server and had the same issue. Although static batching works, it disables Triton's dynamic batching feature.

**Comment by glenn-jocher** [comment]
Hello,

Thank you for sharing your experience with deploying the model on Triton Inference Server. The issue with dynamic batching often stems from how the model handles variable input sizes, which can be particularly challenging with complex architectures.

A potential workaround could be to implement a custom batching logic within your application that groups inference requests into fixed batch sizes before sending them to Triton. This approach allows you to manage batch sizes dynamically on the application side while using static batching on the server side to maintain compatibility.

If you need further assistance or have more details to share, please let us know. We're here to help! 🚀

**Comment by glenn-jocher** [comment]
@laugh12321 The error you're experiencing is related to the attention layer in YOLOv8-worldv2, which has difficulty handling dynamic shapes during TensorRT export. This is a known limitation with attention-based architectures in TensorRT.

For a working solution, I recommend:

1. Use static batch size instead of dynamic:
```
trtexec.exe --onnx=yolov8s-worldv2.onnx --saveEngine=yolov8s-worldv2.engine --workspace=4096
```

2. Increase the workspace size significantly as noted in previous comments (the `--workspace` parameter above specifies MB)

3. If you must use varying batch sizes, consider either:
   - Creating multiple engines with different fixed batch sizes
   - Implementing batch padding in your application code to use a fixed-size engine

This is unfortunately a limitation of the current attention implementation in YOLOv8-worldv2 when exported to TensorRT.
